### PR TITLE
Object Explorer: Refresh OE node if database restored

### PR DESCRIPTION
**Pull request by abist**
Fixes https://github.com/Microsoft/sqlopsstudio/issues/381

### CODE REVIEW

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts [hunk 1]
```diff
@@ -153,9 +155,15 @@ export class RestoreDialogController implements IRestoreDialogController {
 			restoreOption.taskExecutionMode = TaskExecutionMode.executeAndScript;
 		}
 
-		this._restoreService.restore(this._ownerUri, restoreOption);
-		let restoreDialog = this._restoreDialogs[this._currentProvider];
-		restoreDialog.close();
+		this._restoreService.restore(this._ownerUri, restoreOption).then(() => {
```
Comment (llali) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: restore creates a tasks so when the promise response is back, it doesn't mean the database is restored. it mean the task is created. you need to do this in the task manager looking for restore tasks and when they are complete, refresh  the tree node

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts [hunk 2]
```diff
@@ -153,9 +155,15 @@ export class RestoreDialogController implements IRestoreDialogController {
...
+			let connectionProfile = this._connectionService.getConnectionProfile(this._ownerUri);
+			let activeNode = this._objectExplorerService.getObjectExplorerNode(connectionProfile);
+			this._objectExplorerService.refreshTreeNode(activeNode.getSession(), activeNode).then(() => {
+				this._objectExplorerService.getServerTreeView().refreshTree();
```
Comment (llali) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: check serverTreeView is not unidentified

Comment (llali) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: are you refreshing the whole tree? if the connection is a server connection, then you need to refresh databases folder. I think you should handle part of this in the service

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts [hunk 3]
```diff
@@ -153,9 +155,15 @@ export class RestoreDialogController implements IRestoreDialogController {
 			restoreOption.taskExecutionMode = TaskExecutionMode.executeAndScript;
 		}
 
-		this._restoreService.restore(this._ownerUri, restoreOption);
-		let restoreDialog = this._restoreDialogs[this._currentProvider];
-		restoreDialog.close();
+		this._restoreService.restore(this._ownerUri, restoreOption).then(() => {
+			let connectionProfile = this._connectionService.getConnectionProfile(this._ownerUri);
+			let activeNode = this._objectExplorerService.getObjectExplorerNode(connectionProfile);
+			this._objectExplorerService.refreshTreeNode(activeNode.getSession(), activeNode).then(() => {
```
Comment (MattIrv) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: It's possible to have a dashboard open for a connection that doesn't have a corresponding OE node (for example by opening a new connection with "None" as the server group), so we should check if `activeNode` is undefined before using it

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts (hunk 1 above)
Comment (abist) [reply src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: I've rewritten this part to use the taskService now.

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts [hunk 4]
```diff
@@ -153,9 +160,29 @@ export class RestoreDialogController implements IRestoreDialogController {
...
 		}
 
-		this._restoreService.restore(this._ownerUri, restoreOption);
-		let restoreDialog = this._restoreDialogs[this._currentProvider];
-		restoreDialog.close();
+		this._restoreService.restore(this._ownerUri, restoreOption).then(() => {
+			const self = this;
+			let connectionProfile = self._connectionService.getConnectionProfile(self._ownerUri);
+			let activeNode = self._objectExplorerService.getObjectExplorerNode(connectionProfile);
+			this._taskService.onTaskComplete(response => {
```
Comment (MattIrv) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: It looks like this code will add this handler every time a restore action is run, but the handler doesn't do any sort of check to verify that the completed task is the restore task. That means that if you restore a database on some connection, then restore a database on a different connection, then after the second restore this code will execute again and you'll refresh both the first and second connections.

That's probably not a huge issue, but the intention is for it only to run for the task that was just kicked off, right? In that case I think some information about the task is returned by the call to `restoreService.restore` above that you might be able to use

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts [hunk 5]
```diff
@@ -153,9 +160,29 @@ export class RestoreDialogController implements IRestoreDialogController {
...
+				if (this.isSuccessfulRestore(response) && activeNode) {
+					self._objectExplorerService.refreshTreeNode(activeNode.getSession(), activeNode).then(result => {
+						self._objectExplorerService.getServerTreeView().refreshTree();
```
Comment (MattIrv) [inline src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: Why do we need to call `refreshTree` here? Does the call to `refreshTreeNode` just update the backing model but not actually show the changes in the view?

It seems unlikely that anyone would ever want to call `refreshTreeNode` on the object explorer service and not have the refresh actually show up in the view, so maybe it's best to just add this call to the `refreshTreeNode` code inside of OE service?

File: src/sql/parts/registeredServer/viewlet/serverTreeView.ts [hunk 6]
```diff
@@ -441,4 +441,11 @@ export class ServerTreeView {
 		this._tree.dispose();
 		this._toDispose = dispose(this._toDispose);
 	}
+
+	/**
+	 * Return the ITree associated
+	 */
+	public getTree(): ITree {
+		return this._tree;
+	}
```
Comment (MattIrv) [inline src/sql/parts/registeredServer/viewlet/serverTreeView.ts]: I don't see this used anywhere

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts (hunk 5 above)
Comment (abist) [reply src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: No refreshTreeNode doesn't update the view. I've checked the OE code as well, the ITree has to be refreshed for the updated view.

**Comment by abist** [comment]
@AbbiePetcht As far as I know, there can be other tasks other than the restore task. If I add the part for refreshing the OE node in the TaskService, it won't be an independent TaskService in my opinion. I think we should have the underlying work done after a task is completed in their respective areas, and keep the task implementation as an independent library.

**Comment by AbbiePetcht** [comment]
@abist ok it's fine for now. We can revisit this if we think refresh OE is generic enough to put in task service.

File: src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts (hunk 5 above)
Comment (AbbiePetcht) [reply src/sql/parts/disasterRecovery/restore/common/restoreServiceImpl.ts]: Is refreshTree() trying to refresh the whole tree?

---
In reply to: [166411781](https://github.com/Microsoft/sqlopsstudio/pull/607#discussion_r166411781) [](ancestors = 166411781)

**Pull request merged**